modules/cluster_estimation/cluster_estimation.py

```diff
@@ -0,0 +1,365 @@
+        """
+        assert (class_private_create_key is ClusterEstimation.__create_key), "Use create() method"
+
+        # Initializes VGMM
+        self.__vgmm = sklearn.mixture.BayesianGaussianMixture(
+            covariance_type=self.__COVAR_TYPE,
+            n_components=self.__MAX_NUM_COMPONENTS,
+            init_params=self.__MODEL_INIT_PARAM,
+            weight_concentration_prior=self.__WEIGHT_CONCENTRATION_PRIOR,
+            mean_precision_prior=self.__MEAN_PRECISION_PRIOR,
+            max_iter=self.__MAX_MODEL_ITERATIONS,
+            random_state=random_state,
+        )
+
+        # Points storage
+        self.__all_points = []
+        self.__current_bucket = []
+
+        # Requirements to decide to run
+        self.__min_activation_threshold = min_activation_threshold
+        self.__min_new_points_to_run = min_new_points_to_run
+        self.__has_ran_once = False
+
+    def run(self, detections: "list[detection_in_world.DetectionInWorld]", run_override: bool) \
+        -> "tuple[bool, list[object_in_world.ObjectInWorld] | None]":
+        """
+        Take in list of landing pad detections and return list of estimated landing pad locations
+        if number of detections is sufficient, or if manually forced to run.
+
+        PARAMETERS
+        ----------
+        detections: list[DetectionInWorld]
+            List containing DetectionInWorld objects which holds real-world positioning data to run
+            clustering on.
+        
+        run_override: bool
+            Forces ClusterEstimation to predict if data is available, regardless of any other
+            requirements.
+
+        RETURNS
+        -------
+        model_ran: bool
+            True if ClusterEstimation object ran its estimation model.
+
+        objects_in_world: list[ObjectInWorld] or None.
+            List containing ObjectInWorld objects, containing position and covariance value.
+            None if conditions not met and model not ran.
+        """
+        # Store new input data
+        self.__current_bucket += self.__convert_detections_to_point(detections)
+
+        # Decide to run
+        if not self.__decide_to_run(run_override):
+            return False, None
+
+        # Fit points and get cluster data
+        self.__vgmm = self.__vgmm.fit(self.__all_points)
+        model_output: "list[tuple[np.ndarray, float, float]]" = list(
+            zip(
+                self.__vgmm.means_,
+                self.__vgmm.weights_,
+                self.__vgmm.covariances_,
+            )
+        )
+
+        # Empty cluster removal
+        model_output = self.__filter_by_points_ownership(model_output)
+
+        # Sort weights from largest to smallest, along with corresponding covariances and means
+        model_output = self.__sort_by_weights(model_output)
+
+        # Filter out all clusters after __WEIGHT_DROP_THRESHOLD weight drop occurs
+        viable_clusters = [model_output[0]]
```

Comment:
huh. thats interesting, the weight drop threshold doesnt actually do much, since unused clusters get removed by the self.__filter_by_points_ownnership() and bad estimates gets removed by self.__filter_by_covariances.